src/Pagination/Paginator.php

```diff
@@ -0,0 +1,187 @@
+        $this->pageSize = $pageSize;
+    }
+
+    public function paginate(int $page = 1): self
+    {
+        $this->currentPage = max(1, $page);
+        $firstResult = ($this->currentPage - 1) * $this->pageSize;
+
+        $query = $this->queryBuilder
+            ->setFirstResult($firstResult)
+            ->setMaxResults($this->pageSize)
+            ->getQuery();
+
+        if (0 === \count($this->queryBuilder->getDQLPart('join'))) {
+            $query->setHint(CountWalker::HINT_DISTINCT, false);
+        }
+
+        $paginator = new DoctrinePaginator($query, true);
+
+        $useOutputWalkers = \count($this->queryBuilder->getDQLPart('having') ?: []) > 0;
+        $paginator->setUseOutputWalkers($useOutputWalkers);
+
+        $this->results = $paginator->getIterator();
+        $this->numResults = $paginator->count();
+
+        return $this;
+    }
+
+    public function getCurrentPage(): int
+    {
+        return $this->currentPage;
+    }
+
+    public function getLastPage(): int
+    {
+        return (int) ceil($this->numResults / $this->pageSize);
+    }
+
+    public function getPageSize(): int
+    {
+        return $this->pageSize;
+    }
+
+    public function hasPreviousPage(): bool
+    {
+        return $this->currentPage > 1;
+    }
+
+    public function getPreviousPage(): int
+    {
+        return max(1, $this->currentPage - 1);
+    }
+
+    public function getStartPage(): int
+    {
+        return $this->startPage;
+    }
+
+    public function getEndPage(): int
+    {
+        return $this->endPage;
+    }
+
+    public function hasNextPage(): bool
+    {
+        return $this->currentPage < $this->getLastPage();
+    }
+
+    public function getNextPage(): int
+    {
+        return min($this->getLastPage(), $this->currentPage + 1);
+    }
+
+    public function hasToPaginate(): bool
+    {
+        return $this->numResults > $this->pageSize;
+    }
+
+    public function getNumResults(): int
+    {
+        return $this->numResults;
+    }
+
+    public function getNumberOfPages(): int
+    {
+        $numberOfPages = $this->calculateNumberOfPages();
+
+        if (0 === $numberOfPages) {
+            return 1;
+        }
+
+        return $numberOfPages;
+    }
+
+    private function calculateNumberOfPages(): int
+    {
+        return (int) ceil($this->getNumResults() / self::PAGE_SIZE);
+    }
+
+    public function getResults(): Traversable
+    {
+        return $this->results;
+    }
+
+    public function count(): int
+    {
+        return count($this->getResults());
+    }
+
+    public function getIterator(): Traversable
+    {
+        $results = $this->getResults();
+
+        if ($results instanceof \Iterator) {
+            return $results;
+        }
+
+        if ($results instanceof \IteratorAggregate) {
+            return $results->getIterator();
+        }
+
+        return new \ArrayIterator($results);
+    }
+
+    public function calculateStartAndEndPage(): void
```

Comment:
Should this method be public?

Reply:
Yes. It is used in the Twig extension just before rendering our widget. We can't calculate the start/end page in the constructor of the Paginator because the result set isn't known at that time. It has to be used after the `paginate` call in the controller, because at that point we know the current page, number of results, etc.. . Pagerfanta also uses a similar method in their Twig extension to solve this issue.

Reply:
ok.